### PR TITLE
feat(bloom-filter): integrate indexer with mito2 

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#5176 

## What's changed and what's your intention?

Integration with mito2, tested in `test_bloom_filter_indexer`

Some miscellaneous:
* remove `src/mito2/src/sst/index/inverted_index/creator/temp_provider.rs` and ports to `src/mito2/src/sst/index/intermediate.rs`
* `src/mito2/src/sst/index/inverted_index/codec.rs` -> `src/mito2/src/sst/index/codec.rs`

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [ ] API changes are backward compatible.
- [ ] Schema or data changes are backward compatible.

### CODE REVIEW

File: src/datatypes/src/schema.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ use snafu::{ensure, ResultExt};
 use crate::error::{self, DuplicateColumnSnafu, Error, ProjectArrowSchemaSnafu, Result};
 use crate::prelude::ConcreteDataType;
 pub use crate::schema::column_schema::{
-    ColumnSchema, FulltextAnalyzer, FulltextOptions, Metadata, SkippingIndexOptions,
+    ColumnSchema, FulltextAnalyzer, FulltextOptions, Metadata, SkipIndexType, SkippingIndexOptions,
```
Comment (waynexia) [inline src/datatypes/src/schema.rs:31]: It's SKIPPING index

File: src/mito2/src/sst/index.rs [hunk 2]
```diff
@@ -278,6 +318,53 @@ impl<'a> IndexerBuilder<'a> {
...
+
+        if !create {
+            debug!(
+                "Skip creating bloom filter due to config, region_id: {}, file_id: {}",
+                self.metadata.region_id, self.file_id,
+            );
+            return None;
+        }
+
+        let mem_limit = Some(100 * 1024 * 1024); // TODO(zhongzc): add config for bloom filter
```
Comment (evenyag) [inline src/mito2/src/sst/index.rs:333]: The default memory limit may be too large. What about `16M`?

File: src/mito2/src/sst/index.rs [hunk 3]
```diff
@@ -278,6 +318,53 @@ impl<'a> IndexerBuilder<'a> {
...
+                return indexer;
+            }
+            Err(err) => err,
+        };
+
+        if cfg!(any(test, feature = "test")) {
+            panic!(
+                "Failed to create bloom filter, region_id: {}, file_id: {}, err: {}",
+                self.metadata.region_id, self.file_id, err
+            );
```
Comment (evenyag) [inline src/mito2/src/sst/index.rs:358]: non-blocking: Consider using debug format to print the error.
```suggestion
            panic!(
                "Failed to create bloom filter, region_id: {}, file_id: {}, err: {:?}",
                self.metadata.region_id, self.file_id, err
            );
```

File: src/mito2/src/sst/index/indexer/abort.rs [hunk 4]
```diff
@@ -64,4 +65,25 @@ impl Indexer {
...
+        };
+        let Err(err) = indexer.abort().await else {
+            return;
+        };
+
+        if cfg!(any(test, feature = "test")) {
+            panic!(
+                "Failed to abort bloom filter, region_id: {}, file_id: {}, err: {}",
+                self.region_id, self.file_id, err
+            );
```
Comment (evenyag) [inline src/mito2/src/sst/index/indexer/abort.rs:81]:
```suggestion
            panic!(
                "Failed to abort bloom filter, region_id: {}, file_id: {}, err: {:?}",
                self.region_id, self.file_id, err
            );
```

File: src/mito2/src/sst/index/indexer/finish.rs [hunk 5]
```diff
@@ -169,6 +178,43 @@ impl Indexer {
...
+                return true;
+            }
+            Err(err) => err,
+        };
+
+        if cfg!(any(test, feature = "test")) {
+            panic!(
+                "Failed to finish bloom filter, region_id: {}, file_id: {}, err: {}",
+                self.region_id, self.file_id, err
+            );
```
Comment (evenyag) [inline src/mito2/src/sst/index/indexer/finish.rs:207]:
```suggestion
            panic!(
                "Failed to finish bloom filter, region_id: {}, file_id: {}, err: {:?}",
                self.region_id, self.file_id, err
            );
```

File: src/mito2/src/sst/index/indexer/update.rs [hunk 6]
```diff
@@ -80,4 +83,29 @@ impl Indexer {
...
+
+        let Err(err) = creator.update(batch).await else {
+            return true;
+        };
+
+        if cfg!(any(test, feature = "test")) {
+            panic!(
+                "Failed to update bloom filter, region_id: {}, file_id: {}, err: {}",
+                self.region_id, self.file_id, err
+            );
```
Comment (evenyag) [inline src/mito2/src/sst/index/indexer/update.rs:101]:
```suggestion
            panic!(
                "Failed to update bloom filter, region_id: {}, file_id: {}, err: {:?}",
                self.region_id, self.file_id, err
            );
```

File: src/mito2/src/sst/index/bloom_filter/creator.rs [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+        let global_memory_usage = Arc::new(AtomicUsize::new(0));
+
+        for column in &metadata.column_metadatas {
+            let options =
+                column
+                    .column_schema
+                    .skipping_index_options()
+                    .context(IndexOptionsSnafu {
+                        column_name: &column.column_schema.name,
+                    })?;
```
Comment (evenyag) [inline src/mito2/src/sst/index/bloom_filter/creator.rs:91]: non-blocking: Maybe we should store the deserialized index options in the column metadata in the future.

File: src/mito2/src/sst/index/bloom_filter/creator.rs [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+
+        if self.creators.is_empty() {
+            return Ok(());
+        }
+
+        if let Err(update_err) = self.do_update(batch).await {
+            // clean up garbage if failed to update
+            if let Err(err) = self.do_cleanup().await {
+                if cfg!(any(test, feature = "test")) {
+                    panic!("Failed to clean up index creator, err: {err}",);
```
Comment (evenyag) [inline src/mito2/src/sst/index/bloom_filter/creator.rs:138]: We may need to use debug format to get the root cause
```suggestion
                    panic!("Failed to clean up index creator, err: {err:?}",);
```

**Review by evenyag — APPROVED** [review]

**Review by waynexia — APPROVED** [review]

**Pull request merged**